src/types/router.rs

```diff
@@ -0,0 +1,366 @@
+//! The core of the router library.
+//!
+//! The engine module builds a graph given an input of nodes. Path
+//! finding algorithms are also provided to find the shortest path
+//! between two nodes.
+
+/// The router engine module.
+pub mod engine {
+    use std::collections::HashMap;
+
+    use ordered_float::OrderedFloat;
+    use petgraph::{algo::astar, graph::NodeIndex, stable_graph::StableDiGraph};
+
+    use crate::{
+        types::node::{AsNode, Node},
+        utils::graph::build_edges,
+    };
+
+    /// A Router struct contains a graph of nodes and also a hashmap
+    /// that maps a node to its index in the graph.
+    #[derive(Debug)]
+    pub struct Router<'a> {
+        graph: StableDiGraph<&'a Node, OrderedFloat<f32>>,
+        node_indeces: HashMap<&'a Node, NodeIndex>,
+    }
+
+    /// Path finding algorithms.
+    #[derive(Debug, Copy, Clone)]
+    pub enum Algorithm {
+        /// The Dikstra algorithm.
+        Dijkstra,
+        /// The A Star algorithm.
+        AStar,
+    }
+
+    impl Router<'_> {
+        /// Creates a new router with the given graph.
+        pub fn new(
+            nodes: &[impl AsNode],
+            constraint: f32,
+            constraint_function: fn(&dyn AsNode, &dyn AsNode) -> f32,
+            cost_function: fn(&dyn AsNode, &dyn AsNode) -> f32,
+        ) -> Router {
+            println!("[1/4] Initializing the router engine...");
+            println!("[2/4] Building edges...");
+
+            let edges = build_edges(nodes, constraint, constraint_function, cost_function);
+            let mut node_indeces = HashMap::new();
+            let mut graph = StableDiGraph::new();
+
+            println!("[3/4] Building the graph...");
+            for edge in edges {
+                let from_index = *node_indeces
+                    .entry(edge.0)
```

Comment:
If edge vectors were structs instead, this would be more readable and less error-prone. `edge.from_node` rather than `edge.0`. Likewise `edge.2` would be more readable as `edge.cost`.

Reply:
[7673e3d](https://github.com/Arrow-air/lib-router/pull/7/commits/7673e3d63ce4739409501a8fb83b974bf69dc838)